sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/IncrementalExecution.scala

```diff
@@ -216,63 +222,140 @@ class IncrementalExecution(
+      Some(watermarkPropagator.getInputWatermarkForLateEvents(currentBatchId,
+        stateInfo.operatorId))
+    }
+
+    private def inputWatermarkForEviction(stateInfo: StatefulOperatorStateInfo): Option[Long] = {
+      Some(watermarkPropagator.getInputWatermarkForEviction(currentBatchId, stateInfo.operatorId))
+    }
+
+    override def apply(plan: SparkPlan): SparkPlan = {
+      simulateWatermarkPropagation(plan)
+      plan transform {
+        case s: StateStoreSaveExec if s.stateInfo.isDefined =>
+          s.copy(
+            eventTimeWatermarkForLateEvents = inputWatermarkForLateEvents(s.stateInfo.get),
+            eventTimeWatermarkForEviction = inputWatermarkForEviction(s.stateInfo.get)
+          )
+
+        case s: SessionWindowStateStoreSaveExec if s.stateInfo.isDefined =>
+          s.copy(
+            eventTimeWatermarkForLateEvents = inputWatermarkForLateEvents(s.stateInfo.get),
+            eventTimeWatermarkForEviction = inputWatermarkForEviction(s.stateInfo.get)
+          )
+
+        case s: SessionWindowStateStoreRestoreExec if s.stateInfo.isDefined =>
+          s.copy(
+            eventTimeWatermarkForLateEvents = inputWatermarkForLateEvents(s.stateInfo.get),
+            eventTimeWatermarkForEviction = inputWatermarkForEviction(s.stateInfo.get)
+          )
+
+        case s: StreamingDeduplicateExec if s.stateInfo.isDefined =>
+          s.copy(
+            eventTimeWatermarkForLateEvents = inputWatermarkForLateEvents(s.stateInfo.get),
+            eventTimeWatermarkForEviction = inputWatermarkForEviction(s.stateInfo.get)
+          )
+
+        case m: FlatMapGroupsWithStateExec if m.stateInfo.isDefined =>
+          m.copy(
+            eventTimeWatermarkForLateEvents = inputWatermarkForLateEvents(m.stateInfo.get),
+            eventTimeWatermarkForEviction = inputWatermarkForEviction(m.stateInfo.get)
+          )
+
+        case m: FlatMapGroupsInPandasWithStateExec if m.stateInfo.isDefined =>
+          m.copy(
+            eventTimeWatermarkForLateEvents = inputWatermarkForLateEvents(m.stateInfo.get),
+            eventTimeWatermarkForEviction = inputWatermarkForEviction(m.stateInfo.get)
+          )
+
+        case j: StreamingSymmetricHashJoinExec =>
+          val iwLateEvents = inputWatermarkForLateEvents(j.stateInfo.get)
+          val iwEviction = inputWatermarkForEviction(j.stateInfo.get)
+          j.copy(
+            eventTimeWatermarkForLateEvents = iwLateEvents,
+            eventTimeWatermarkForEviction = iwEviction,
+            stateWatermarkPredicates =
+              StreamingSymmetricHashJoinHelper.getStateWatermarkPredicates(
+                j.left.output, j.right.output, j.leftKeys, j.rightKeys, j.condition.full,
+                iwEviction, !allowMultipleStatefulOperators)
+          )
+      }
     }
   }
 
-  override def preparations: Seq[Rule[SparkPlan]] = state +: super.preparations
+  override def preparations: Seq[Rule[SparkPlan]] = Seq(
+    shufflePartitionsRule,
```

Comment:
The actual reason is that issuing state op ID needs to be performed "before" simulate propagation. (That said, it needs to traverse the entire tree at least 2 times.) After that, I feel like rule for issuing deals with so many things by itself compared to the propagation, hence refactor out to similar level of rules. Yeah you're right some of them are for refactoring and it may trigger 2 more passes. 

I'll see whether we can apply pruning with the node pattern, and compose them into two if there is no good way to do that.

Reply:
No pattern seems to be registered for "physical" node. Do you know whether this is somewhat discouraged, or it's just that there was no reason to do so?

https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/trees/TreePatterns.scala

If that's the latter, I'll add some for refactored rules. Otherwise I'll try to consolidate rules. Might be great if we could deal with follow-up ticket/PR.

Reply:
I just addressed this. Please refer to the change of [c39b642](https://github.com/apache/spark/pull/39931/commits/c39b64284b1298c358fa559ec1b91d0d2f7739a5).